tests/scene/test_visual_shader.h

```diff
@@ -0,0 +1,202 @@
+		CHECK(vs->get_node_position(VisualShader::TYPE_COLLIDE, 3) == Vector2(1, 1));
+
+		Ref<VisualShaderNode> vsn2 = memnew(VisualShaderNodeInput);
+		CHECK(vsn2.is_valid());
+		vs->add_node(VisualShader::TYPE_FOG, vsn2, Vector2(1, 2), 4);
+		CHECK(vs->get_node_position(VisualShader::TYPE_FOG, 4) == Vector2(1, 2));
+		vs->set_node_position(VisualShader::TYPE_FOG, 4, Vector2(2, 2));
+		CHECK(vs->get_node_position(VisualShader::TYPE_FOG, 4) == Vector2(2, 2));
+	}
+
+	SUBCASE("Testing VisualShaderNode ID") {
+		Ref<VisualShader> vs = memnew(VisualShader);
+		CHECK(vs.is_valid());
+
+		Ref<VisualShaderNode> vsn1 = memnew(VisualShaderNodeInput);
+		CHECK(vsn1.is_valid());
+		vs->add_node(VisualShader::TYPE_COLLIDE, vsn1, Vector2(0, 0), 3);
+		CHECK(vs->get_valid_node_id(VisualShader::TYPE_COLLIDE) - 1 == 3);
+
+		Ref<VisualShaderNode> vsn2 = memnew(VisualShaderNodeInput);
+		CHECK(vsn2.is_valid());
+		vs->add_node(VisualShader::TYPE_FOG, vsn2, Vector2(1, 2), 4);
+		CHECK(vs->get_valid_node_id(VisualShader::TYPE_FOG) - 1 == 4);
+
+		Ref<VisualShaderNode> vsn3 = memnew(VisualShaderNodeInput);
+		CHECK(vsn3.is_valid());
+		vs->add_node(VisualShader::TYPE_LIGHT, vsn3, Vector2(4, 2), 5);
+		CHECK(vs->get_valid_node_id(VisualShader::TYPE_LIGHT) - 1 == 5);
+
+		Ref<VisualShaderNode> vsn4 = memnew(VisualShaderNodeInput);
+		CHECK(vsn4.is_valid());
+		vs->add_node(VisualShader::TYPE_SKY, vsn4, Vector2(1, 10), 6);
+		CHECK(vs->get_valid_node_id(VisualShader::TYPE_SKY) - 1 == 6);
+	}
+
+	SUBCASE("Testing remove and replace VisualShaderNode") {
+		Ref<VisualShader> vs = memnew(VisualShader);
+		CHECK(vs.is_valid());
+
+		ERR_PRINT_OFF;
+
+		Ref<VisualShaderNode> vsn1 = memnew(VisualShaderNodeInput);
+		CHECK(vsn1.is_valid());
+		vs->add_node(VisualShader::TYPE_COLLIDE, vsn1, Vector2(0, 0), 3);
+		CHECK(vs->get_node(VisualShader::TYPE_COLLIDE, 3) == vsn1);
+		vs->remove_node(VisualShader::TYPE_COLLIDE, 3);
+		CHECK_FALSE(vs->get_node(VisualShader::TYPE_COLLIDE, 3) == vsn1);
+
+		Ref<VisualShaderNode> vsn2 = memnew(VisualShaderNodeInput);
+		CHECK(vsn2.is_valid());
+		vs->add_node(VisualShader::TYPE_FOG, vsn2, Vector2(1, 2), 4);
+		CHECK(vs->get_node(VisualShader::TYPE_FOG, 4) == vsn2);
+		vs->remove_node(VisualShader::TYPE_FOG, 4);
+		CHECK_FALSE(vs->get_node(VisualShader::TYPE_FOG, 4) == vsn2);
+
+		Ref<VisualShaderNode> vsn3 = memnew(VisualShaderNodeInput);
+		CHECK(vsn3.is_valid());
+		vs->add_node(VisualShader::TYPE_LIGHT, vsn3, Vector2(4, 2), 5);
+		CHECK(vs->get_node(VisualShader::TYPE_LIGHT, 5) == vsn3);
+		vs->remove_node(VisualShader::TYPE_LIGHT, 5);
+		CHECK_FALSE(vs->get_node(VisualShader::TYPE_LIGHT, 5) == vsn3);
+
+		Ref<VisualShaderNode> vsn4 = memnew(VisualShaderNodeInput);
+		CHECK(vsn4.is_valid());
+		vs->add_node(VisualShader::TYPE_SKY, vsn4, Vector2(1, 10), 6);
+		CHECK(vs->get_node(VisualShader::TYPE_SKY, 6) == vsn4);
+		vs->remove_node(VisualShader::TYPE_SKY, 6);
+		CHECK_FALSE(vs->get_node(VisualShader::TYPE_SKY, 6) == vsn4);
```

Comment:
Similar, this can be rewritten using loops.

Reply:
Done 👍